web/application.html

```diff
@@ -0,0 +1,209 @@
+            } else {
+                if (document.exitFullscreen) document.exitFullscreen();
+                else if (document.webkitExitFullscreen) document.webkitExitFullscreen();
+                else if (document.mozCancelFullScreen) document.mozCancelFullScreen();
+                else if (document.msExitFullscreen) document.msExitFullscreen();
+            }
+        }
+
+        function isFullScreen() {
+            return !!(document.fullscreenElement || document.webkitFullscreenElement || document.mozFullScreenElement || document.msFullscreenElement);
+        }
+
+        // Emscripten triggered ready event when the app is actually started
+        function ready() {
+            document.getElementById('canvas').style.display = 'block';
+
+            // Small timeout for notifying engine with the actual canvas size
+            setTimeout(() => {
+                Module.JSCanvasSize(canvasWidth, canvasHeight, false, devicePixelRatio);
+            }, 100);
+        }
+
+        var Module = {
+            preRun: [],
+            postRun: [],
+            binary: false,
+            info: false,
+            ready: false,
+            receiver: false,
+            rendererSize: {
+                width: canvasWidth * devicePixelRatio,
+                height: canvasHeight * devicePixelRatio,
+                aspectRatio: canvasWidth / canvasHeight,
+            },
+            canvas: canvasElement,
+
+            print: function (text) {
+                console.log(text);
+            },
+
+            printErr: function(text) {
+                console.error(text);
+            },
+
+            // Urho3D called method which tells the canvas the current renderer resolution
+            SetRendererSize: function(width, height) {
+                console.log('Engine renderer size changed to', width, height);
+                this.rendererSize.width  = width;
+                this.rendererSize.height = height;
+                this.rendererSize.aspectRatio = width / height;
+
+                if (isFullScreen()) {
+                    return;
+                }
+
+                if (this.canvas.width === width && this.canvas.height === height) {
+                    return;
+                }
+                console.log('Updating canvas size, width', this.canvas.width, 'to', width, 'and height', this.canvas.height, 'to', height);
+                this.canvas.width = width;
+                this.canvas.height = height;
+                this.canvas.style.height = (canvasWidth / Module.rendererSize.aspectRatio) + 'px';
+            },
+
+            // Retrieve the current status of the application
+            setStatus: function(text) {
+                if (text === 'Running...') {
+                    ready();
+                }
+                if (Module.setStatus.interval) clearInterval(Module.setStatus.interval);
+                var m = text.match(/([^(]+)\((\d+(\.\d+)?)\/(\d+)\)/);
```

Comment:
I don't know JS best practices, but formatting data to the string to decode it later feels hacky.
Why function param cannot be used instead?
Is it fixed-signature function or something?

Reply:
This part is actually comes from one of the emscripten shells, haven't change it so far.